_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,421 @@
+        assertEq(sub(token.balanceOf(dst), preBalDst), delta);
+    }
+
+}
+```
+
+We can execute this test by running `dapp test --rpc-url <URL>`, and we see that a counter example
+is found: the balancer token reverts for transfers to the zero address.
+
+```
+[FAIL] prove_transfer(address,uint256)
+
+Failure: prove_transfer(address,uint256)
+
+  Counter Example:
+
+    result:   Revert
+    calldata: prove_transfer(0x0000000000000000000000000000000000000000, 0)
+```
+
+## Interactive Exploration
+
+`hevm` also includes a visual debugger, and we can use this to interactively explore the execution
+tree:
+
+<INSERT_ASCINEMA_HERE>
+
+## Limitations, Assumptions & Future Work
+
+**Non Linearity (`safeMul`)**
+
+The symbolic execution engine in `hevm` is backed by an [SMT
+solver](https://en.wikipedia.org/wiki/Satisfiability_Modulo_Theories) (currently either `z3` or
+`cvc4` are supported). Expressions involving non-linear arithmetic (multiplication or divison by
+symbolic variables) are extremely challenging for SMT solvers, and it will often not be possible to
+symbolically execute tests involving lots of non linear arithmetic.
+
+Unfortunately, non linear arithmetic is quite common in real world contracts (e.g.
+[`safeMul`](https://github.com/dapphub/ds-math/blob/master/src/math.sol#L25) can easily
+involve both a multiplication and a division by a symbolic variable).
+
+We hope to include optimizations in future releases of `hevm` that reduce the load on the solver
+when executing contracts that make use of common non-linearities (`safeMul` included).
+
+**Symbolic Representation of Dynamic Types**
+
+`hevm` is currently unable to represent dynamic types (e.g. `bytes`, `string`) symbolically. Tests
+that need symbolic representations of dynamic data will currently fail with an `Unsupported symbolic
+abiencoding` error.
+
+We intend to lift this restriction in a future release of `hevm`.
+
+**Symbolic Constructor Arguments**
+
+Contract bytecode is currently assumed by `hevm` to be completely concrete. As constructor arguments
+are implemented on the evm level by appending data to the contract's `creationCode`, symbolic
+execution of contract constructors where the arguments are set to symbolic values will currently
+fail with an `UnexpectedSymbolicArg` error.
+
+We intend to lift this restriction in a future release of `hevm`.
+
+**State Explosion**
+
+Symbolic execution explores all possible paths through a program. If the program is large, or
+contains many branches, this can become computationly very intensive. This issue is known as "state
+explosion". While the relative simplicity of most smart contracts limits the impact, you should be
+aware that exploration of very large or complex contract systems may become very time consuming.
+
+In these cases it be convenient to first write your tests as fuzz tests, and only begin symbolically
+executing once you have a set of properties that you are happy with.
+
+**Storage Collisions**
```

Comment:
this may be too much information, not sure

Reply:
it's a pretty reasonable assumption to make, and maybe not something ppl need to be worried about?